### ISSUE TITLE
SVG overlay per slide plugin - discussion, questions

### ISSUE THREAD

**Issue by boskar** [openseadragon/openseadragon#732]
As I plan to create svg overlay per slide plugin I've got some technical questions in order to make it right.

I'm going to use existing addOverlay and clearOverlays as I've proven that it works and scales right ( http://wrong.technology/~boskar/histo-www/ ) 

1. Which one would be better:
One big SVG cointaining multiple elements (in groups?) in one full slide overlay

Potentially we could add everything we could ever need in the svg tree, extending (yay _X_ml!) 

PROS:
Entire complex overlay system in one file, easily loaded on slide change
CONS: 
Issue with point overlays - We will have to mark them somehow and make the pointer scalable in different way that the rest of picture
Will the browser handle "terapixel-size" svg tree easily? ,
Some browsers may not support direct interaction with svg elements (I had problems with tooltips even in IE9 o maybe it was even early IE10 - not important actually)
Potentially we break (extend) svg standard

We might need to parse the SVG XML to get additional info (let's say - tooltips content)?

--OR--
Multiple svg - each overlay added by seperate addOverlay call 

PROS:
Proven to work correctly 
We can interact with the overlay at higher level in a tree 
(the svg element itself - CSS styles or even with div on higher level)
We can permit user to use updateOverlay
Small SVGs can be made by any editor, and used keeping the SVG untouched. 

CONS:
Complex 
It would require definition file/query with informations such as URLs to SVGs and their respective locations in the viewer. 

I'm totally against any SVG manipulating plugins - so far I've managed to achieve everything using ordinary jquery. 
I do not intend to draw from scratch - style changes, hiding elements or blinking should be enough, and jquery is enough here.

**Comment by avandecreme** [comment]
>As I plan to create svg overlay per slide plugin

What would that plugin do?

**Comment by dwrogers** [comment]
Option 2 seems more flexible to me.

“I'm totally against any SVG manipulating plugins - so far I've managed to achieve everything using ordinary jquery.
I do not intend to draw from scratch - style changes, hiding elements or blinking should be enough, and jquery is enough here.”

Primarily out if ignorance, I have to ask why you are against having plugins manipulating SVGs since I will need to draw from scratch for my application. I was thinking about creating separate plugins for the various artifacts I need for a pathology slide viewer: user-drawn regions of interest, user-drawn annotations (perhaps those two can be merged into a single Plugin/SVG layer), and various user-manipulatable tools such as measurement and area-bound counting constructs. It seems like it would be much more complex to have all of those in a single layer. In some cases, one might have layers which need to scale, but also have another layer or two which do not scale. Just brainstorming here.

David
--
David W. Rogers
Software Developer / Collaborative Data Services / 206.667.7089 / drogers@fredhutch.org / Fred Hutch / Cures Start Here

**Comment by boskar** [comment]
>> As I plan to create svg overlay per slide plugin

> What would that plugin do?

Load set of overlays corresponding to what is opened by the OSD viewer, as it is opened and destroyed as it is closed.

**Comment by boskar** [comment]
@dwrogers 
Well, yes, You're obviously right. 
As soon as we start to draw from scratch SVG manipulation plugins start to make sense. 
At the moment the idea was more simple - static tiles, static overlays, easy to use, easy to save and create server side. 

But at the moment approach presented in https://github.com/openseadragon/svg-overlay/ looks far from what I need in my project, oriented on teaching - I've got the svg generated on the server, maybe I'll even make it static. (BTW, the overlays are created in Pannoramic Viewer, extracted directly from the mrxs, and translated to svg by custom script). 

Moreover `this._node.setAttribute('transform','translate(' + p.x + ',' + p.y + ') scale(' + scale + ')');` is a working solution of scaling, but I've found another, imho simplier, without any plugins, using svg viewBox.

**Comment by boskar** [comment]
and some questions:

1) What's the right ® configuration for tooltips on overlays? 

Currently I'm using :     
$("polygon[title]").tooltip({ placement: 'auto top', container: 'body', viewport: {selector: '#view', padding: 2 } });

but the tooltips sometimes appear when the overlay is not visible in the OSD viewport. Maybe I should change the container?

2) How to get current tilesource URL? I'm planing to use it as identifier for what is opened currently, as I don't have any other idea. 

The flowchart is: 
1. New slide loaded (via open/page change)
2. open/page event fired (passing the URL to binded function?)
3. Overlays definition file (json/xml/svg/whatever) XHRed with tilesource URL as GET parameter
4. Overlays drawn

**Comment by boskar** [comment]
Oh, and there is another, third option. 

It seems wrong from the beginning: 
https://msdn.microsoft.com/en-us/library/cc645048(v=vs.95).aspx

Only Rectangles supported, but, for compatibility, we could try to first: fully support and second: extend the original DZI specification. 
Storing overlays in DZI *is* an option, but as we support different input formats, and DZI is quite often (ok, at least in my case) generated on the fly by server side program, it sounds unreasonable.

**Comment by avandecreme** [comment]
From what I understand, I wouldn't do it that way.

I probably would have an array of slides like this:
```javascript
var slides = [{
   tileSource: "slide1.dzi", // or whatever your tile source is
   overlays: [overlay1, overlay2...] //the overlays object can be either URLs to fetch or JSON
}, {
   tileSource: "slide1.dzi", // or whatever your tile source is
   overlays: [overlay1, overlay2...] //the overlays object can be either URLs to fetch or JSON
}];
```
With that, you can get the tiles sources array like this to pass them to OSD:
```javascript
var tileSources = slides.map(function(slide) {
   return slide.tileSource;
});
var viewer = new OpenSeadragon({
   tileSources: tileSources
});
```

Then you can register to the page event to retrieve the overlays:
```javascript
viewer.addHandler("page", function(event) {
    var page = event.page;
    var overlays = slides[page].overlays;
});
```

**Comment by boskar** [comment]
Yes, It is a good idea too, for some what You've posted is even sufficient as It is. 

On the other hand - it is not as easy as it could be, from user's point of view. 
Ideally the event handler and the second code block should be hidden (in a plugin), and all the user should enter is the array.

Ideally - open function / tilesource parameter should do (makes sense only if we put overlays in dzi like in the third solution)  

And, in case of not having every slide in an order ("open" function) I'd be a bit afraid of relying on page index. 

We still need some sort of format definiton.

**Comment by iangilman** [comment]
For what it's worth, option 2 seems like the way to go to me as well. On the other hand, I'd be curious to know if there is a way the openseadragon-svg-overlay plugin could be used for your scenario. 

As for getting the URL to the current tile source, I don't know that that information is easily accessible...it's a fairly common request, though, so maybe we should file an issue. Here is the line where that information is lost: 

https://github.com/openseadragon/openseadragon/blob/master/src/tilesource.js#L431

At any rate, because of that, the sort of structure @avandecreme suggests is the best. You just need to tuck that structure away inside of your plugin so your users don't need to be aware of it.

**Comment by boskar** [comment]
So let's try. 
What about:
> 1) What's the right ® configuration for tooltips on overlays?
> Currently I'm using :
> $("polygon[title]").tooltip({ placement: 'auto top', container: 'body', viewport: {selector: '#view', padding: 2 } });
> but the tooltips sometimes appear when the overlay is not visible in the OSD viewport. Maybe I should change the container?

**Comment by iangilman** [comment]
Looks like you're using jQuery UI tooltips? I don't know enough about them to give a recommendation. Have you tried getting the overlay element directly and then applying a tooltip to it?

**Comment by boskar** [comment]
No, it's bootstrap/jquery at the moment. Jqueryui was used in the very beginning of the project. The jqueryui code is commented out. 
The overlay element from OSD? It is rectangular, do adjacent irregular elements overlap, giving bad results.

**Comment by boskar** [comment]
And the actual problem is where to put the tooltip element itself. body and #view work, but it could be better. I'll show the issue as I come back from the holidays.

**Comment by iangilman** [comment]
Hmm...yeah, doing hit testing on irregular overlays for a Bootstrap tooltip could require a bit of finesse. Is it possible to manually trigger the tooltip via JavaScript? Perhaps you could listen to OSD for mousemove events (via a MouseTracker) and do the hit calculation yourself and then pass it onto Bootstrap.

**Comment by boskar** [comment]
Yes, it's possible to trigger it manually. 
At the moment triggering tooltips from hover on overlays works surprisingly good, I've worked it out by myself. 

But I've got some other ideas/problems to use tooltips
1. Hover on overlay -> scroll the viewer with keyboard -> overlay does not point to the item.
2. I wanted to add persistent overlays mode (show all of them at the beginning and keep them) - They've added correctly, but they don't scroll. 

Possible solution: Bind to animation event and redraw them.
drawback: It will be fired *after* the animation, as long as the user moves the viewport tooltips will be in wrong location.
solution: destroy them on the animation-start and draw on animation-finish.
drawback: no tooltips during animation... acceptable.

But maybe there is other solution? I want them to stick tooltips to overlays.
The tooltip is a div, a direct child of body. There might be an element which moves as the tiles move, where I could add the tooltips? (Canvas?) But I want them to scroll, but not to zoom. 

I could probably achieve it if I make the tooltips completely from scratch and add them as point overlay in OSD, but this way I lose "the magic" with autopositioning, keeping them visible etc... 

It becomes more and more complex.

3. Issue: Some tooltips are shown despite their overlay is not in the viewport (if triggered manually). 
Well, maybe I'll just prevent triggering them unless overlay is visible...

**Comment by iangilman** [comment]
Moving them via the viewer's animation event should be fine; that's what the svg-overlay plugin does, and it's pretty solid. 

As for your other questions, it's getting hard for me to visualize...I'd have to see it in action.
